klaus/contrib/auto_klaus.py

```diff
@@ -0,0 +1,103 @@
+"""
+Alternative take on the "automatically discovered repositories" concept
+that requires no threads, polling or inotify. Instead the filesystem is
+consulted whenever a repository name is looked up.
+
+Since os.path.exists() and os.listdir() are fairly quick filesystem
+operations, performance should be good for small to medium sites.
+FancyRepo() objects are cached.
+
+Repositories are identified by the existence of a
+
+    <reponame>/git-daemon-export-ok
+
+file (for compatibility with gitweb).
+
+Example usage:
+
+    from klaus.contrib.auto_klaus import Klaus, SlashDynamicRepos
+
+    application = Klaus('/srv/git', "My git repositories", False)
+
+    application.wsgi_app = httpauth.AlwaysFailingAuthMiddleware(
+        wsgi_app=dulwich.web.make_wsgi_chain(
+            backend=dulwich.server.DictBackend(SlashDynamicRepos(application.valid_repos)),
+            fallback_app=application.wsgi_app,
+        ),
+    )
+"""
+
+import collections.abc
+import pathlib
+import os
+import os.path
+
+import klaus
+import klaus.repo
+
+class FilesystemRepoDict(collections.abc.Mapping):
+    """
+    Maintain a virtual read-only dictionary whose contents represent
+    the presence of git repositories in the given root directory.
+    """
+
+    def __init__(self, root):
+        self._root = pathlib.Path(root)
+        self._repos = {}
+
+    def __getitem__(self, name):
+        if not name or name[0] == '.' or '/' in name or '\0' in name:
+            raise KeyError(name)
+
+        repos = self._repos
+        path = self._root / name
+        if not os.path.exists(path / 'git-daemon-export-ok'):
```

Comment:
what about non-bare repositories?

Reply:
I suppose we could make the path to check configurable, so you can configure it to be `.git/git-daemon-export-ok` or even something completely different if you don't care about gitweb conventions.

Reply:
Checking for git-daemon-export-ok is quite specific to being compatible with gitweb, and a change from the behaviour that currently exists in klaus. For example, most of my repositories that are visible in klaus don't have a git-daemon-export-ok file and wouldn't work with this script. Perhaps it would make sense to call that out more explicitly?

Reply:
I agree that it's kind of annoying that we have multiple ways to deal with autodiscovery/autoreloading. It long overdue someone contributes a well thought out solution into Klaus proper.

I wonder if we should make this "flag file" approach part of the existing autoreloading code? To me it seems the only user-visible difference between this implementation and the existing one the `git-daemon-export-ok` file.